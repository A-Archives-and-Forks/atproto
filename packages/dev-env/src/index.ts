export * from './bsky'
export * from './network'
export * from './network-no-appview'
export * from './pds'
export * from './plc'
export * from './feed-gen'
export * from './moderator-client'
export * from './seed'
export * from './types'
export * from './util'
