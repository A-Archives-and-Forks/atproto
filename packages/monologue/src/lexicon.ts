import { schemas as bskySchemas } from '@atproto/api'
import { Infer } from '@atproto/jetstream'
import { LexiconDoc } from '@atproto/lexicon'

export const schemas = [
  ...bskySchemas,

  {
    lexicon: 1,
    id: 'chat.bsky.monologue.defs',
    defs: {
      monologueViewSubject: {
        type: 'object',
        required: ['did'],
        properties: {
          did: { type: 'string', format: 'did' },
        },
      },
      monologueView: {
        type: 'object',
        required: ['subject', 'muted', 'unreadCount'],
        properties: {
          subject: {
            type: 'union',
            refs: [
              'lex:app.bsky.actor.defs#profileViewBasic',
              'lex:chat.bsky.monologue.defs#monologueViewSubject',
            ],
          },
          muted: { type: 'boolean' },
          unreadCount: { type: 'integer' },
        },
      },
      messageViewAuthor: {
        type: 'object',
        required: ['did'],
        properties: {
          did: { type: 'string', format: 'did' },
        },
      },
      messageView: {
        type: 'object',
        required: ['id', 'author', 'timestamp', 'text'],
        properties: {
          id: { type: 'string', format: 'at-uri' },
          author: {
            type: 'ref',
            ref: 'lex:chat.bsky.monologue.defs#messageViewAuthor',
          },
          timestamp: { type: 'string', format: 'datetime' },

          text: {
            type: 'string',
            maxLength: 10000,
            maxGraphemes: 1000,
          },
          facets: {
            type: 'array',
            description: 'Annotations of text (mentions, URLs, hashtags, etc)',
            items: { type: 'ref', ref: 'lex:app.bsky.richtext.facet#main' },
          },
          embed: {
            type: 'union',
            refs: ['lex:app.bsky.embed.record#view'],
          },
        },
      },
      deletedMessageView: {
        type: 'object',
        required: ['id', 'author', 'timestamp'],
        properties: {
          id: { type: 'string', format: 'at-uri' },
          author: {
            type: 'ref',
            ref: 'lex:chat.bsky.monologue.defs#messageViewAuthor',
          },
          timestamp: { type: 'string', format: 'datetime' },
        },
      },
    },
  },

  {
    lexicon: 1,
    id: 'chat.bsky.monologue.deleteMessage',
    defs: {
      main: {
        type: 'procedure',
        input: {
          encoding: 'application/json',
          schema: {
            type: 'object',
            required: ['messageId'],
            properties: {
              messageId: { type: 'string', format: 'cid' },
            },
          },
        },
      },
    },
  },
  {
    lexicon: 1,
    id: 'chat.bsky.monologue.getMessages',
    defs: {
      main: {
        type: 'query',
        parameters: {
          type: 'params',
          required: ['subject'],
          properties: {
            subject: { type: 'string', format: 'did' },
            limit: {
              type: 'integer',
              minimum: 1,
              maximum: 100,
              default: 50,
            },
            cursor: { type: 'string', format: 'at-uri' },
          },
        },
        output: {
          encoding: 'application/json',
          schema: {
            type: 'object',
            required: ['messages'],
            properties: {
              cursor: { type: 'string', format: 'at-uri' },
              messages: {
                type: 'array',
                items: {
                  type: 'union',
                  refs: [
                    'lex:chat.bsky.monologue.defs#messageView',
                    'lex:chat.bsky.monologue.defs#deletedMessageView',
                  ],
                },
              },
            },
          },
        },
      },
    },
  },

  {
    lexicon: 1,
    id: 'chat.bsky.monologue.list',
    defs: {
      main: {
        type: 'query',
        parameters: {
          type: 'params',
          properties: {
            limit: {
              type: 'integer',
              minimum: 1,
              maximum: 100,
              default: 50,
            },
            cursor: { type: 'string' },
          },
        },
        output: {
          encoding: 'application/json',
          schema: {
            type: 'object',
            required: ['monologues'],
            properties: {
              cursor: { type: 'string' },
              monologues: {
                type: 'array',
                items: {
                  type: 'ref',
                  ref: 'lex:chat.bsky.monologue.defs#monologueView',
                },
              },
            },
          },
        },
      },
    },
  },
  {
    lexicon: 1,
    id: 'chat.bsky.monologue.message',
    defs: {
      main: {
        type: 'record',
        description: 'A message in a monologue.',
        record: {
          type: 'object',
          required: ['subject', 'text'],
          properties: {
            subject: { type: 'string', format: 'did' },
            createdAt: { type: 'string', format: 'datetime' },
            text: {
              type: 'string',
              maxLength: 10000,
              maxGraphemes: 1000,
            },
            facets: {
              type: 'array',
              description:
                'Annotations of text (mentions, URLs, hashtags, etc)',
              items: { type: 'ref', ref: 'lex:app.bsky.richtext.facet' },
            },
            embed: {
              type: 'union',
              refs: ['lex:app.bsky.embed.record#view'],
            },
          },
        },
      },
    },
  },

  {
    lexicon: 1,
    id: 'chat.bsky.monologue.mute',
    defs: {
      main: {
        type: 'procedure',
        input: {
          encoding: 'application/json',
          schema: {
            type: 'object',
            required: ['monologueId'],
            properties: {
              monologueId: { type: 'string', format: 'did' },
            },
          },
        },
      },
    },
  },

  {
    lexicon: 1,
    id: 'chat.bsky.monologue.unmute',
    defs: {
      main: {
        type: 'procedure',
        input: {
          encoding: 'application/json',
          schema: {
            type: 'object',
            required: ['monologueId'],
            properties: {
              monologueId: { type: 'string', format: 'did' },
            },
          },
        },
      },
    },
  },

  {
    lexicon: 1,
    id: 'chat.bsky.monologue.updateRead',
    defs: {
      main: {
        type: 'procedure',
        input: {
          encoding: 'application/json',
          schema: {
            type: 'object',
            required: ['monologueId'],
            properties: {
              monologueId: { type: 'string', format: 'did' },
              rev: { type: 'string' },
            },
          },
        },
        output: {
          encoding: 'application/json',
          schema: {
            type: 'object',
            required: ['monologue'],
            properties: {
              monologue: {
                type: 'ref',
                ref: 'lex:chat.bsky.monologue.defs#monologueView',
              },
            },
          },
        },
      },
    },
  },
] as const satisfies LexiconDoc[]

export type Schemas = typeof schemas

export type I<T extends string> = Infer<Schemas, T>
