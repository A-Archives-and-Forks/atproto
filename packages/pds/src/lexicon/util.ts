/**
 * GENERATED CODE - DO NOT MODIFY
 */
export type $Type<Id extends string, Hash extends string> = Hash extends 'main'
  ? Id | `${Id}#${Hash}`
  : `${Id}#${Hash}`

function has$type<V>(v: V): v is V & object & { $type: unknown } {
  return v != null && typeof v === 'object' && '$type' in v
}

function check$type<Id extends string, Hash extends string>(
  $type: unknown,
  id: Id,
  hash: Hash,
): $type is $Type<Id, Hash> {
  return (
    typeof $type === 'string' &&
    ($type === id
      ? hash === 'main'
      : // $type === `${id}#${hash}`
        $type.length === id.length + 1 + hash.length &&
        $type[id.length] === '#' &&
        $type.startsWith(id) &&
        $type.endsWith(hash))
  )
}

export function is$typed<V, Id extends string, Hash extends string>(
  v: V,
  id: Id,
  hash: Hash,
): v is V & object & { $type: $Type<Id, Hash> } {
  return has$type(v) && check$type(v.$type, id, hash)
}
