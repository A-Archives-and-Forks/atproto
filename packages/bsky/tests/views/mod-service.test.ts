import AtpAgent, { AtUri } from '@atproto/api'
import { TestNetwork, SeedClient, basicSeed, RecordRef } from '@atproto/dev-env'
import { forSnapshot, stripViewerFromModService } from '../_util'
import { ids } from '../../src/lexicon/lexicons'

describe('mod service views', () => {
  let network: TestNetwork
  let agent: AtpAgent
  let pdsAgent: AtpAgent
  let sc: SeedClient

  // account dids, for convenience
  let alice: string
  let bob: string

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'bsky_views_mod_service',
    })
    agent = network.bsky.getClient()
    pdsAgent = network.pds.getClient()
    sc = network.getSeedClient()
    await basicSeed(sc)
    alice = sc.dids.alice
    bob = sc.dids.bob

    const aliceRes = await pdsAgent.api.com.atproto.repo.createRecord(
      {
        repo: alice,
        collection: ids.AppBskyModerationService,
        rkey: 'self',
        record: {
          description: 'alices labeler',
          policies: {
            description: 'long policy description',
            reportReasons: [
              'com.atproto.moderation.defs#reasonSpam',
              'com.atproto.moderation.defs#reasonOther',
            ],
            labelValues: ['spam', '!hide', 'scam', 'impersonation'],
          },
          createdAt: new Date().toISOString(),
        },
      },
      { headers: sc.getHeaders(alice), encoding: 'application/json' },
    )
    await pdsAgent.api.com.atproto.repo.createRecord(
      {
        repo: bob,
        collection: ids.AppBskyModerationService,
        rkey: 'self',
        record: {
          description: 'bobs labeler',
          policies: {
            description: 'another policy description',
            reportReasons: ['com.atproto.moderation.defs#reasonSexual'],
            labelValues: ['nudity', 'sexual', 'porn'],
          },
          createdAt: new Date().toISOString(),
        },
      },
      { headers: sc.getHeaders(bob), encoding: 'application/json' },
    )

    await sc.like(bob, new RecordRef(aliceRes.data.uri, aliceRes.data.cid))
    await network.processAll()
  })

  afterAll(async () => {
    await network.close()
  })

  it('fetches mod service', async () => {
    const view = await agent.api.app.bsky.moderation.getService(
      { did: alice },
      { headers: await network.serviceHeaders(bob) },
    )

    expect(forSnapshot(view.data)).toMatchSnapshot()
  })

  it('fetches multiple mod services', async () => {
    const view = await agent.api.app.bsky.moderation.getServices(
      { dids: [alice, bob, 'did:example:missing'] },
      { headers: await network.serviceHeaders(bob) },
    )

    expect(forSnapshot(view.data)).toMatchSnapshot()
  })

  it('fetches mod service unauthed', async () => {
    const { data: authed } = await agent.api.app.bsky.moderation.getService(
      { did: alice },
      { headers: await network.serviceHeaders(bob) },
    )
    const { data: unauthed } = await agent.api.app.bsky.moderation.getService({
      did: alice,
    })
    expect(unauthed.view).toEqual(stripViewerFromModService(authed.view))
  })

  it('fetches multiple mod services unauthed', async () => {
    const { data: authed } = await agent.api.app.bsky.moderation.getServices(
      {
        dids: [alice, bob, 'did:example:missing'],
      },
      { headers: await network.serviceHeaders(bob) },
    )
    const { data: unauthed } = await agent.api.app.bsky.moderation.getServices({
      dids: [alice, bob, 'did:example:missing'],
    })
    expect(unauthed.views.length).toBeGreaterThan(0)
    expect(unauthed.views).toEqual(authed.views.map(stripViewerFromModService))
  })

  it('blocked by mod service takedown', async () => {
    const uri = AtUri.make(alice, ids.AppBskyModerationService, 'self')
    await network.bsky.ctx.dataplane.updateTakedown({
      recordUri: uri.toString(),
      takenDown: true,
    })
    const promise = agent.api.app.bsky.moderation.getService(
      { did: alice },
      { headers: await network.serviceHeaders(bob) },
    )

    await expect(promise).rejects.toThrow('could not find moderation service')

    const res = await agent.api.app.bsky.moderation.getServices(
      { dids: [alice, bob] },
      { headers: await network.serviceHeaders(bob) },
    )
    expect(res.data.views.length).toBe(1)
    expect(res.data.views[0].creator.did).toEqual(bob)

    // Cleanup
    await network.bsky.ctx.dataplane.updateTakedown({
      actorDid: alice,
      takenDown: false,
    })
  })
})
