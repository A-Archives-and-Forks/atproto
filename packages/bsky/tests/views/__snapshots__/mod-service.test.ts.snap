// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mod service views fetches mod service 1`] = `
Object {
  "cid": "cids(0)",
  "creator": Object {
    "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
    "description": "its me!",
    "did": "user(0)",
    "displayName": "ali",
    "handle": "alice.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [
      Object {
        "cid": "cids(2)",
        "cts": "1970-01-01T00:00:00.000Z",
        "neg": false,
        "src": "user(0)",
        "uri": "record(3)",
        "val": "self-label-a",
      },
      Object {
        "cid": "cids(2)",
        "cts": "1970-01-01T00:00:00.000Z",
        "neg": false,
        "src": "user(0)",
        "uri": "record(3)",
        "val": "self-label-b",
      },
    ],
    "viewer": Object {
      "blockedBy": false,
      "followedBy": "record(2)",
      "following": "record(1)",
      "muted": false,
    },
  },
  "description": "alices labeler",
  "indexedAt": "1970-01-01T00:00:00.000Z",
  "labels": Array [],
  "likeCount": 1,
  "policies": Object {
    "description": "long policy description",
    "labelValues": Array [
      "spam",
      "!hide",
      "scam",
      "impersonation",
    ],
    "reportReasons": Array [
      "com.atproto.moderation.defs#reasonSpam",
      "com.atproto.moderation.defs#reasonOther",
    ],
  },
  "uri": "record(0)",
  "viewer": Object {
    "like": "record(4)",
  },
}
`;

exports[`mod service views fetches multiple mod services 1`] = `
Object {
  "views": Array [
    Object {
      "cid": "cids(0)",
      "creator": Object {
        "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
        "description": "its me!",
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "labels": Array [
          Object {
            "cid": "cids(2)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(3)",
            "val": "self-label-a",
          },
          Object {
            "cid": "cids(2)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(3)",
            "val": "self-label-b",
          },
        ],
        "viewer": Object {
          "blockedBy": false,
          "followedBy": "record(2)",
          "following": "record(1)",
          "muted": false,
        },
      },
      "description": "alices labeler",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "likeCount": 1,
      "uri": "record(0)",
      "viewer": Object {
        "like": "record(4)",
      },
    },
    Object {
      "cid": "cids(3)",
      "creator": Object {
        "avatar": "https://bsky.public.url/img/avatar/plain/user(3)/cids(1)@jpeg",
        "description": "hi im bob label_me",
        "did": "user(2)",
        "displayName": "bobby",
        "handle": "bob.test",
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "muted": false,
        },
      },
      "description": "bobs labeler",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "likeCount": 0,
      "uri": "record(5)",
      "viewer": Object {},
    },
  ],
}
`;
